src/planetary_system.jl

```diff
@@ -352,6 +352,340 @@ function generate_period_and_sizes_power_law(s::Star, sim_param::SimParam; num_p
 end
 
 
+
+
+
+using SpecialFunctions
+
+cdf_lognormal(x::Float64; μ::Float64=0., σ::Float64=1.) = 0.5*erfc(-(log(x) - μ)/(σ*sqrt(2.)))
+
+function invert_cdf_lognormal(y::Float64; μ::Float64=0., σ::Float64=1.)
+    @assert 0<=y<=1
+    return exp(μ - sqrt(2.)*σ*erfcinv(2*y))
+end
+
+function cdf_power_law(x::Float64, x0::Float64, x1::Float64, α::Float64)
+    @assert x0 <= x <= x1
+    if α != -1
+        return (x^(α+1) - x0^(α+1))/(x1^(α+1) - x0^(α+1))
+    else # if α == -1
+        return log(x/x0)/log(x1/x0)
+    end
+end
+
+function invert_cdf_power_law(y::Float64, x0::Float64, x1::Float64, α::Float64)
+    @assert x0 < x1
+    if α != -1
+        return (x0^(α+1) + (x1^(α+1) - x0^(α+1))*y)^(1/(α+1))
+    else # if α == -1
+        return x0*(x1/x0)^y
+    end
+end
+
+function draw_segmented_uniform(segments::Array{Tuple{Float64,Float64},1})
+    # Note: "segments" must be non-overlapping
+    n_segs = length(segments) # number of segments
+    if n_segs == 0
+        println("No segments given; returning NaN.")
+        return NaN
+    end
+    for (i,seg) in enumerate(segments)
+        @assert 0 <= seg[1] < seg[2] <= 1 # check that all segments are valid
+        if i != n_segs
+            @assert segments[i][2] < segments[i+1][1] # check to make sure there are no overlapping segments
+        end
+    end
+
+    seg_lengths = [seg[2]-seg[1] for seg in segments]
+    sum_lengths = sum(seg_lengths)
+
+    u_draw = rand()
+    for (i,seg) in enumerate(segments)
+        # Map the union of the segments uniformly to the domain [0,1]:
+        if sum(seg_lengths[1:i-1])/sum_lengths <= u_draw < sum(seg_lengths[1:i])/sum_lengths
```

Comment:
I'd suggest multiplying by sum_lengths, so that we avoid risk of dividing by a small number.
Also, you could compute sum(seg_lengths[1:i-1]) for both limits and then just add seg_lengths[i] to get the upper limit

Reply:
Thanks, I've updated the function to implement these points.